### PR TITLE
Fix performance regression with _InheritedTheme

### LINKED ISSUE #10250: Theme/IconTheme regression
https://github.com/flutter/flutter/pull/10179 caused a regression to our build times

regression1

### PR DESCRIPTION
**Pull request by xster**
#10250

Confirmed in tracer but I'm not sure the unit test is right.

### CODE REVIEW

File: packages/flutter/test/material/theme_test.dart [hunk 1]
```diff
@@ -265,4 +265,40 @@ void main() {
...
+class Test extends StatefulWidget {
+  @override
+  _TestState createState() => new _TestState();
+}
+
+class _TestState extends State<Test> {
+
+  @override
+  Widget build(BuildContext context) {
+    ++testBuildCalled;
```
Comment (Hixie) [inline packages/flutter/test/material/theme_test.dart]: fwiw, and you should feel free to ignore me here if you disagree, i find `testBuildCalled += 1` reads slightly better than using the `++` prefix operator. (then again, i'd remove the `++` operator from the language entirely if i could!)

File: packages/flutter/test/material/theme_test.dart [hunk 2]
```diff
@@ -265,4 +265,40 @@ void main() {
...
+}
+
+int testBuildCalled;
+class Test extends StatefulWidget {
+  @override
+  _TestState createState() => new _TestState();
+}
+
+class _TestState extends State<Test> {
+
```
Comment (Hixie) [inline packages/flutter/test/material/theme_test.dart]: nit: extraneous blank line

File: packages/flutter/test/material/theme_test.dart [hunk 3]
```diff
@@ -265,4 +265,40 @@ void main() {
...
+      final ThemeData themeData = new ThemeData(primaryColor: const Color(0xFF000000));
+
+      await tester.pumpWidget(
+        new MaterialApp(
+          theme: themeData,
+          home: new Test(),
+        ),
+      );
+      expect(testBuildCalled, 1);
+      tester.element(find.byType(MaterialApp)).markNeedsBuild();
```
Comment (Hixie) [inline packages/flutter/test/material/theme_test.dart]: I would not have expected this to trigger the bug even without your fix. The child doesn't change, so the system should short-circuit it and not rebuild it, and the child doesn't call Theme.of(), so it shouldn't get artificially rebuilt when the Theme is rebuilt.

I think a better test would be to just have an explicit Theme widget (no MaterialApp) with a child that is a `const Test()` whose build calls `Theme.of`, which you pumpWidget twice (with the same exact tree and the same instance of themeData). The Test should only be built once. Then you pumpWidget again but with a new instance of ThemeData with the same values. Test should still not be built again. Then you pumpWidget a fourth time, same tree, but with a slightly different instance of ThemeData. Now finally Test should rebuild.

**Comment by Hixie** [comment]
Nice work tracking down this bug! I'm really glad it was an actual but. I was really worried it was fundamental to the framework design! :-)

File: packages/flutter/test/material/theme_test.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/test/material/theme_test.dart]: woohoo, that was exactly the advice I was looking for. Was reading the Element code but couldn't figure out when rebuilds are triggered. Thanks!

Confirmed your suggested test fails at the previous commit point

File: packages/flutter/test/material/theme_test.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/test/material/theme_test.dart]: I don't have a strong opinion so done

**Comment by Hixie** [comment]
New test looks great!
LGTM

**Comment by Hixie** [comment]
@xster Is there anything that needs to happen before you land this?

**Comment by xster** [comment]
Ummm not sure why I didn't submit. Will merge on green.

**Pull request merged**